modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/NonFungibleErc1155JsonBalanceFetcher.swift

```diff
@@ -95,56 +95,28 @@ public class NonFungibleErc1155JsonBalanceFetcher {
+        }.map(on: queue, { [tokensService] ercTokens in
+            tokensService.addCustom(tokens: ercTokens, shouldUpdateBalance: false)
 
             return contractsAndTokenIds
         })
     }
 
     private func fetchUnknownErc1155ContractsDetails(contractsAndTokenIds: Erc1155TokenIds.ContractsAndTokenIds) -> Promise<[ERCToken]> {
-        let contractsToAdd: [AlphaWallet.Address] = contractsAndTokenIds.keys.filter { contract in
-            tokensService.token(for: contract, server: server) == nil
-        }
-        guard !contractsToAdd.isEmpty else { return Promise<[ERCToken]>.value(.init()) }
-        let (promise, seal) = Promise<[ERCToken]>.pending()
-        //Can't use `DispatchGroup` because `ContractDataDetector.fetch()` doesn't call `completion` once and only once
-        var contractsProcessed: Set<AlphaWallet.Address> = .init()
-        var erc1155TokensToAdd: [ERCToken] = .init()
-        func markContractProcessed(_ contract: AlphaWallet.Address) {
-            contractsProcessed.insert(contract)
-            if contractsProcessed.count == contractsToAdd.count {
-                seal.fulfill(erc1155TokensToAdd)
-            }
-        }
-        for each in contractsToAdd {
-            ContractDataDetector(address: each, account: account, server: server, assetDefinitionStore: assetDefinitionStore, analytics: analytics).fetch { data in
-                switch data {
-                case .name, .symbol, .balance, .decimals:
-                    break
-                case .nonFungibleTokenComplete(let name, let symbol, let balance, let tokenType):
-                    let token = ERCToken(
-                            contract: each,
-                            server: self.server,
-                            name: name,
-                            symbol: symbol,
-                            //Doesn't matter for ERC1155 since it's not used at the token level
-                            decimals: 0,
-                            type: tokenType,
-                            balance: balance
-                    )
-                    erc1155TokensToAdd.append(token)
-                    markContractProcessed(each)
-                case .fungibleTokenComplete:
-                    markContractProcessed(each)
-                case .delegateTokenComplete:
-                    markContractProcessed(each)
-                case .failed:
-                    //TODO we are ignoring `.failed` here because it is called multiple times and we need to wait until `ContractDataDetector.fetch()`'s `completion` is called once and only once
-                    break
+        let contractsToAdd: [AlphaWallet.Address] = contractsAndTokenIds.keys.filter { tokensService.token(for: $0, server: session.server) == nil }
+        let promises = contractsToAdd.map { importToken.fetchTokenOrContract(for: $0, server: session.server) }
+
+        return when(resolved: promises).map(on: queue, { result -> [ERCToken] in
```

Comment:
I think this changes the semantic; before PR, if there are any failures the `fetchUnknownErc1155ContractsDetails()` wouldn't resolve the promise it returns. But it does with the new version. Looks better. Great!